File: affirm/src/main/java/com/affirm/android/AffirmHttpRequest.java
package com.affirm.android;

import java.util.HashMap;
import java.util.Map;

public final class AffirmHttpRequest {
    private final String url;
    private final Method method;
    private final AffirmHttpBody body;
    private final String tag;
    private final Map<String, String> headers;

    private AffirmHttpRequest(Builder builder) {
        url = builder.mUrl;
        method = builder.mMethod;
        body = builder.mBody;
        tag = builder.mTag;
        headers = builder.headers;
    }

    String getUrl() {
        return url;
    }

    Method getMethod() {
        return method;
    }


    Map<String, String> getAllHeaders() {
        return headers;
    }


    AffirmHttpBody getBody() {
        return body;
    }

    String getTag() {
        return tag;
    }

    public enum Method {
        GET, POST, PUT, DELETE
    }

    static final class Builder {

        private String mUrl;
        private Method mMethod;
        private AffirmHttpBody mBody;
        private String mTag;
        private Map<String, String> headers;

        Builder() {
        }

        public Builder setUrl(String url) {
            mUrl = url;
            return this;
        }


        Builder setMethod(AffirmHttpRequest.Method method) {
            mMethod = method;
            return this;
        }

        Builder setBody(AffirmHttpBody body) {
            mBody = body;
            return this;
        }

        Builder setTag(String tag) {
            mTag = tag;
            return this;
        }

        public Builder setHeaders(Map<String, String> headers) {
            this.headers = new HashMap<>(headers);
            return this;
        }

        AffirmHttpRequest build() {
            return new AffirmHttpRequest(this);
        }
    }
}

File: affirm/src/main/java/com/affirm/android/AffirmHttpRequest.kt
package com.affirm.android

import android.webkit.CookieManager
import java.io.OutputStream
import java.io.UnsupportedEncodingException

internal open class AffirmHttpRequest internal constructor(
    val method: Method,
    val url: String,
    private val body: String? = null
) {
    internal val headers: Map<String, String>
        get() {
            val cookieManager = CookieManager.getInstance()
            val cookie = cookieManager
                .getCookie(AffirmConstants.HTTPS_PROTOCOL + AffirmPlugins.get().baseUrl())
            return mapOf(
                "Accept" to "application/json",
                "Content-Type" to "application/json",
                "Affirm-User-Agent" to "Affirm-Android-SDK",
                "Affirm-User-Agent-Version" to BuildConfig.VERSION_NAME
            )
                .plus(
                    cookie?.let {
                        mapOf("Cookie" to it)
                    }.orEmpty()
                )
        }

    internal open fun writeBody(outputStream: OutputStream) {
        try {
            body?.toByteArray(Charsets.UTF_8).let {
                outputStream.write(it)
                outputStream.flush()
            }
        } catch (e: UnsupportedEncodingException) {
            throw e
        }
    }

    override fun toString(): String {
        return "${method.code} $url"
    }

    internal enum class Method(val code: String) {
        GET("GET"),
        POST("POST")
    }

    internal companion object {
        fun createGet(url: String): AffirmHttpRequest {
            return AffirmHttpRequest(Method.GET, url, null)
        }

        fun createPost(url: String, body: String): AffirmHttpRequest {
            return AffirmHttpRequest(Method.POST, url, body)
        }
    }
}
